demo/faas/code/rust/client/src/lib.rs

```diff
@@ -0,0 +1,92 @@
+use std::str;
+
+use proxy_wasm::{
+    dispatcher, get_buffer, invoke_service, log, set_buffer,
+    traits::{Context, HttpContext, RootContext},
+    types::*,
+};
+
+fn get_query_param(body: &str, param_name: &str) -> Option<String> {
+    for item in body.split("&") {
+        let key = format!("{:}=", param_name);
+        if item.starts_with(&key) {
+            return Some((&item[key.len()..]).to_owned());
+        }
+    }
+    None
+}
+
+fn get_book_name(body: &Option<Bytes>) -> Option<String> {
+    match body {
+        Some(body) => match str::from_utf8(body) {
+            Ok(body) => get_query_param(body, "name"),
+            _ => None,
+        },
+        None => None,
+    }
+}
+
+#[no_mangle]
+pub fn _start() {
+    dispatcher::set_root_context(|_| -> Box<dyn RootContext> { Box::new(ClientHttpRootContext) });
+}
+
+#[no_mangle]
+#[allow(unused_must_use)]
+pub extern "C" fn proxy_get_id() {
+    set_buffer(BufferType::BufferTypeCallData, 0, b"id_1");
+}
+
+struct ClientHttpRootContext;
+
+impl Context for ClientHttpRootContext {}
+
+impl RootContext for ClientHttpRootContext {
+    fn create_http_context(&self, _context_id: u32) -> Option<Box<dyn HttpContext>> {
+        Some(Box::new(ClientHttpContext {}))
+    }
+
+    fn get_type(&self) -> Option<ContextType> {
+        Some(ContextType::HttpContext)
+    }
+}
```

Comment:
How about moving this to `proxy-wasm` crate, looks like this could be a default `RootContext` implementation.

```rust

#[derive(Default)]
pub struct DefaultRootContext<T> {
    _http_context: PhantomData<T>,
}

impl<T> Context for DefaultRootContext<T> where T: HttpContext {}

impl<T> RootContext for DefaultRootContext<T>
where
    T: HttpContext + Default + 'static,
{
    fn create_http_context(&self, _context_id: u32) -> Option<Box<dyn HttpContext>> {
        Some(Box::new(T::default()))
    }

    fn get_type(&self) -> Option<ContextType> {
        Some(ContextType::HttpContext)
    }
}
```

Reply:
updated, please review ^_^